controllers/replace_failed_process_groups_test.go

```diff
@@ -63,6 +68,639 @@ var _ = Describe("replace_failed_process_groups", func() {
+		var targetPodProcessGroupID fdbv1beta2.ProcessGroupID
+		var node *corev1.Node
+		var adminClient *mock.AdminClient
+		var configMap *corev1.ConfigMap
+		var processMap map[fdbv1beta2.ProcessGroupID][]fdbv1beta2.FoundationDBStatusProcessInfo
+
+		BeforeEach(func() {
+			cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions = []fdbv1beta2.TaintReplacementOption{
+				{
+					Key:               &taintKeyStar,
+					DurationInSeconds: &taintKeyStarDuration,
+				},
+				{
+					Key:               &taintKeyMaintenance,
+					DurationInSeconds: &taintKeyMaintenanceDuration,
+				},
+			}
+			cluster.Spec.AutomationOptions.Replacements.FailureDetectionTimeSeconds = pointer.Int(5)
+			cluster.Spec.AutomationOptions.Replacements.TaintReplacementTimeSeconds = pointer.Int(1)
+			// Update cluster config so that generic reconciliation will work
+			err := k8sClient.Update(ctx.TODO(), cluster)
+			Expect(err).NotTo(HaveOccurred())
+
+			adminClient, err = mock.NewMockAdminClientUncast(cluster, k8sClient)
+
+			Expect(err).NotTo(HaveOccurred())
+			databaseStatus, err := adminClient.GetStatus()
+			Expect(err).NotTo(HaveOccurred())
+			processMap = make(map[fdbv1beta2.ProcessGroupID][]fdbv1beta2.FoundationDBStatusProcessInfo)
+			for _, process := range databaseStatus.Cluster.Processes {
+				processID, ok := process.Locality["process_id"]
+				// if the processID is not set we fall back to the instanceID
+				if !ok {
+					processID = process.Locality["instance_id"]
+				}
+				processMap[fdbv1beta2.ProcessGroupID(processID)] = append(processMap[fdbv1beta2.ProcessGroupID(processID)], process)
+			}
+
+			allPods, err = clusterReconciler.PodLifecycleManager.GetPods(ctx.TODO(), clusterReconciler, cluster, internal.GetPodListOptions(cluster, "", "")...)
+			Expect(err).NotTo(HaveOccurred())
+
+			allPvcs = &corev1.PersistentVolumeClaimList{}
+			err = clusterReconciler.List(ctx.TODO(), allPvcs, internal.GetPodListOptions(cluster, "", "")...)
+			Expect(err).NotTo(HaveOccurred())
+
+			configMap = &corev1.ConfigMap{}
+			err = k8sClient.Get(ctx.TODO(), types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Name + "-config"}, configMap)
+			Expect(err).NotTo(HaveOccurred())
+
+			pods, err := clusterReconciler.PodLifecycleManager.GetPods(ctx.TODO(), clusterReconciler, cluster, internal.GetSinglePodListOptions(cluster, "storage-1")...)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pods)).To(Equal(1))
+
+			podOnTaintedNode = pods[0] // Future: choose a random pod to test
+			node = &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{Name: podOnTaintedNode.Spec.NodeName},
+			}
+			targetPodProcessGroupID = internal.GetProcessGroupIDFromMeta(cluster, podOnTaintedNode.ObjectMeta)
+
+			// Call validateProcessGroups to set processGroupStatus to tainted condition
+			processGroupsStatus, err := validateProcessGroups(ctx.TODO(), clusterReconciler, cluster, &cluster.Status, processMap, configMap, allPods, allPvcs)
```

Comment:
I believe I asked the question already but I don't find the comment: What is the benefit of doing all this setup instead of modifying the cluster's status directly to have the expected input? In most cases we try to keep the unit tests focused and not running other methods from the operator all "integration tests" or more complex unit tests are normally done in the `cluster_controller_tests`.

Reply:
I didn't see this comment except here. :)

Since the taint feature piggy back the checking and setting process group's status in the existing validateProcessGroup() function, I treat `validateProcessGroups()` as the minimum unit. 
Testing the `validateProcessGroups()` is better than testing `UpdateTaintCondition()` because it can cover the taint feature's interaction with the existing function that changes a process group's status.